src/lib/badge/badge.ts

```diff
@@ -0,0 +1,221 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Directive, Input, Renderer2, ElementRef} from '@angular/core';
+import {MatIconRegistry} from '@angular/material/icon';
+import {coerceBooleanProperty} from '@angular/cdk/coercion';
+import {ThemePalette} from '@angular/material/core';
+
+let nextId = 0;
+
+ /** @docs-private */
+export class MatBadgeBase {
+
+  /** Theme for the badge */
+  get matBadgeColor(): ThemePalette { return this._color; }
+  set matBadgeColor(value: ThemePalette) {
+    const colorPalette = value;
+
+    if (colorPalette !== this._color) {
+      if (this._color) {
+        this._renderer.removeClass(this._elementRef.nativeElement, `mat-badge-${this._color}`);
+      }
+      if (colorPalette) {
+        this._renderer.addClass(this._elementRef.nativeElement, `mat-badge-${colorPalette}`);
+      }
+
+      this._color = colorPalette;
+    }
+  }
+  private _color: ThemePalette = 'primary';
+
+  /** Whether the badge should overlap its contents or not */
+  set matBadgeOverlap(val: boolean) {
+    this._overlap = coerceBooleanProperty(val);
+  }
+  get matBadgeOverlap(): boolean {
+    return this._overlap;
+  }
+  private _overlap: boolean = true;
+
+  /** Position the badge should reside; 'above|below before|after' */
+  set matBadgePosition(val: string) {
+    this._position = val;
+    this._isAbove = val.indexOf('below') === -1;
+    this._isAfter = val.indexOf('before') === -1;
+  }
+  get matBadgePosition(): string {
+    return this._position;
+  }
+  private _position: string = 'above after';
+
+  _id: number = nextId++;
+  _isAbove: boolean = true;
+  _isAfter: boolean = true;
+
+  constructor(public _renderer: Renderer2, public _elementRef: ElementRef) {}
+
+  /** Clears the previous badge content */
+  _clearContents(): void {
+    const contents = document.getElementById(`badge-content-${this._id}`);
```

Comment:
Couldn't you take a hold of the badge content using a `ViewChild` or `ContentChild`?

Reply:
I thought that only worked for angular inserted dom. This is manually inserted.

Reply:
This plays into my other comment (https://github.com/angular/material2/pull/7483#discussion_r143226186) that it probably shouldn't be constructing DOM in the first place. IMO it should be fine to have it as a component instead of a directive.

Reply:
Ya, thats what I had but @jelbourn wanted this to be a directive that can decorate components like icons. I totally agree w/ it not constructing dom though, open to suggestions.

Reply:
You could create a separate component and dynamically compile and attach it (for `span` or `svg`, then store it in an array for later removal.